tests/AutoReview/Makefile/MakefileTest.php

```diff
@@ -0,0 +1,305 @@
+<?php
+/**
+ * This code is licensed under the BSD 3-Clause License.
+ *
+ * Copyright (c) 2017, Maks Rafalko
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Tests\AutoReview\Makefile;
+
+use function array_filter;
+use function array_shift;
+use function chdir;
+use function current;
+use function file_get_contents;
+use function getcwd;
+use function implode;
+use PHPUnit\Framework\TestCase;
+use function shell_exec;
+use function sprintf;
+use function strpos;
+use function substr_count;
+
+/**
+ * @internal
+ *
+ * @coversNothing
+ *
+ * @group infra
+ */
+final class MakefileTest extends TestCase
+{
+    private const MAKEFILE_PATH = __DIR__ . '/../../../Makefile';
+
+    /**
+     * @var string
+     */
+    private static $originalCwd;
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function setUpBeforeClass(): void
+    {
+        self::$originalCwd = getcwd();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function tearDown(): void
+    {
+        chdir(self::$originalCwd);
+    }
+
+    public function test_the_default_goal_is_the_help_command(): void
+    {
+        $output = shell_exec(sprintf('make -f %s', self::MAKEFILE_PATH));
```

Comment:
I imagine that for some reason the default goal is as before, but build anything in in all possible combination, which also includes running the very this test. This can go on for a long time. How about `make --dry-run` to make sure the above does not happen?

Reply:
I'm not sure to understand you: the make default goal is `help` since #706. Running `--dry-run` right now would output:

```
echo "\033[33mUsage:\033[0m\n  make TARGET\n\n\033[32m#\n# Commands\n#---------------------------------------------------------------------------\033[0m\n"
fgrep -h "##"  Makefile | fgrep -v fgrep | sed -e 's/\\$//' | sed -e 's/##//' | awk 'BEGIN {FS = ":"}; {printf "\033[33m%s:\033[0m%s\n", $1, $2}'
```

Instead of testing the actual output of the help command (which is the purpose of this test)

Reply:
You can test that it mentions `Usage:`, and then proceed with the current test, or stop right there.

What will be procedure to update this fixture? Will it be better to store it in a file and implement [an automatic updater](https://github.com/infection/infection/blob/master/tests/Console/E2ETest.php#L178)?

Reply:
The idea is to really test the output which includes the alignment (although unfortunately is not easy to test). So I was thinking more of: if the Makefile output changes, then you try the output see if everything is alright and when it is you update the expected string. I don't mind storing it in a separate file, but I'm a bit anxious of people just doing `make > expected-file` which is not desired since you should manually check the output for the colors & alignment

Reply:
If you are concerned about the potential time the command takes, how about adding a very low timeout there? That way if the makefile starts doing something unexpected it will fail fast

Reply:
Couldn't find an easy way to do this :(

Reply:
timeout -k 5 make -f ...

Reply:
>I'm more concerned about running make from within make running the same Makefile. It may cause failures that will be very hard to reason about.

 Yeah that's a valid concern. But I think the timeout + the fact that the `help` command doesn't really execute anything is good enough. With https://github.com/infection/infection/pull/729 I'm also isolating from where those tests are being run which should make it easier to spot if there is an issue